File: amw/src/main/java/interpackage/Command.java
package interpackage;

import model.utils.LiteralParser;
import org.apache.commons.lang3.tuple.Pair;
import org.apache.commons.lang3.tuple.Triple;

import java.util.List;
import java.util.stream.Collectors;

import static model.utils.LiteralParser.*;

public class Command {

	private String id;
	private String name;
	private String description;
	private List<Version> versions;

	public Command ( String id, String name, String description, List<Version> versions ) {
		this.id = id;
		this.name = name;
		this.description = description;
		this.versions = versions;
	}

	public String getId ( ) {
		return id;
	}

	public String getName ( ) {
		return name;
	}

	public String getDescription ( ) {
		return description;
	}

	public List<Version> getVersions ( ) {
		return versions;
	}

	static public Command parse ( String input ) {
		Pair<String, String> pair = splitStructAndList( input );
		String id = getValue( pair.getKey( ), "id" );
		String name = getValue( pair.getKey( ), "name" );
		String description = getValue( pair.getKey( ), "description" );
		List<Version> versions = split( pair.getValue( ) ).stream( ).map( Version::parse )
				.collect( Collectors.toList( ) );
		return new Command( id, name, description, versions );
	}

	@Override
	public Command clone( ) {
		return new Command( id, name, description, versions.stream( ).map( Version::clone )
				.collect( Collectors.toList( ) ) );
	}

	@Override
	public boolean equals( Object obj ) {
		if ( ! ( obj instanceof Command ) ) return false;

		Command cmd = ( Command ) obj;
		return id.equals( cmd.id ) && name.equals( cmd.name ) && description.equals( cmd.description )
				&& versions.equals( cmd.versions );
	}

	@Override
	public String toString( ) {
		return "Id: " + id + ", name: " + name + ", description: " + description + ", versions:" + versions.stream( )
				.map( p -> "\t" + p.getLeft( ) + "\n"
						+ "\t\t" + String.join( "", p.getMiddle( ) ) + "\n"
						+ "\t\t" + p.getRight( ) );
	}

	public static class Version extends Triple<String, List<String>, String> {

		private String id;
		private List<String> requirements;
		private String script;

		public Version( String id, List<String> requirements, String script ) {
			this.id = id;
			this.requirements = requirements;
			this.script = script;
		}

		public String getId ( ) {
			return id;
		}

		public List<String> getRequirements ( ) {
			return requirements;
		}

		public String getScript ( ) {
			return script;
		}

		static public Version parse ( String input ) {
			return new Version( getValue( input, "v_id" ),
					split( split( getValue( input ) ).stream( ).filter( lit -> lit.startsWith( "requirements" ) )
							.findFirst( ).map( LiteralParser::splitStructAndList ).get( ).getValue( ) ),
					getValue( input, "script" ) );
		}

		@Override
		public String getLeft ( ) {
			return getId( );
		}

		@Override
		public List<String> getMiddle ( ) {
			return getRequirements( );
		}

		@Override
		public String getRight ( ) {
			return getScript( );
		}

		@Override
		public Version clone( ) {
			return new Version( id, requirements.stream( ).map( String::new ).collect( Collectors.toList( ) ), script );
		}

		@Override
		public boolean equals( Object obj ) {
			if ( ! ( obj instanceof Version ) ) return false;

			Version vrs = ( Version ) obj;
			return id.equals( vrs.id ) && requirements.equals( vrs.requirements ) && script.equals( vrs.script );
		}
	}
}

File: src/main/kotlin/common/type/Command.kt
package common.type

import common.translation.LiteralParser.getValue
import common.translation.LiteralParser.split
import common.translation.LiteralParser.splitStructAndList

/**
 * Data class to store commands info. A command is an operation executable by an agent
 *
 * @author Paolo Baldini
 */
class Command(val id: String, val name: String, val description: String, val versions: List<Version>): Cloneable {

    public override fun clone() = Command(id, name, description, versions.map { it.clone() }.toList())

    override fun equals(other: Any?): Boolean = other is Command && id == other.id

    override fun toString() = "ID: $id, name: $name, Description: $description, Versions:" + versions
            .joinToString { it.toString() }

    override fun hashCode() = id.hashCode()

    class Version(val id: String, val requirements: List<String>, val script: String): Cloneable {

        public override fun clone() = Version(id, requirements.map { s -> s }.toList(), script)

        override fun equals(other: Any?): Boolean = other is Version && id == other.id

        override fun hashCode() = id.hashCode()

        override fun toString() = "ID: $id; Requirements: $requirements; Script: $script"

        /**
         * Parse a string to extract a Version object
         */
        companion object {
            fun parse(input: String) = Version(getValue(input, "v_id")!!,
                    split(splitStructAndList(split(getValue(input)).first { it.startsWith("requirements") }).second),
	                getValue(input, "script")!!)
        }
    }

    /**
     * Parse a string to extract a Command object
     */
    companion object {
        fun parse(input: String): Command {
            val pair = splitStructAndList(input)
            val id = getValue(pair.first, "id")!!
            val name = getValue(pair.first, "name")!!
            val description = getValue(pair.first, "description")!!
            val versions = split(pair.second).filter { it.isNotBlank() }.map { Version.parse(it) }
            return Command(id, name, description, versions)
        }
    }
}